### PR TITLE
Update moshi in library and sample app and use new way of using sealed types

### LINKED ISSUE #100: Runtime crash with moshi 1.9.x
Hey 👋 

I was trying to update all libs to the latest version on my project and after all the updates, I've started to get a crash on runtime with the serialization of the class `com.pusher.pushnotifications.internal.StartJob`

Not 100% sure but I think the crash comes to the fact that on Moshi 1.9.+ some implicit behaviours changed.

I've found this article and I think it's related to this https://www.zacsweers.dev/a-closer-look-at-moshi-1-9/

But here it is the crash I'm getting:

```
2019-12-06 08:21:49.315 7481-7481/ E/AndroidRuntime: FATAL EXCEPTION: main
    Process: , PID: 7481
    java.lang.RuntimeException: Unable to get provider com.pusher.pushnotifications.internal.PushNotificationsInitProvider: java.lang.IllegalArgumentException: Cannot serialize Kotlin type com.pusher.pushnotifications.internal.StartJob. Reflective serialization of Kotlin classes without using kotlin-reflect has undefined and unexpected behavior. Please use KotlinJsonAdapter from the moshi-kotlin artifact or use code gen from the moshi-kotlin-codegen artifact.
    for class com.pusher.pushnotifications.internal.StartJob
    for class com.pusher.pushnotifications.internal.ServerSyncJob
        at android.app.ActivityThread.installProvider(ActivityThread.java:6988)
        at android.app.ActivityThread.installContentProviders(ActivityThread.java:6528)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6445)
        at android.app.ActivityThread.access$1300(ActivityThread.java:219)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1859)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
     Caused by: java.lang.IllegalArgumentException: Cannot serialize Kotlin type com.pusher.pushnotifications.internal.StartJob. Reflective serialization of Kotlin classes without using kotlin-reflect has undefined and unexpected behavior. Please use KotlinJsonAdapter from the moshi-kotlin artifact or use code gen from the moshi-kotlin-codegen artifact.
    for class com.pusher.pushnotifications.internal.StartJob
    for class com.pusher.pushnotifications.internal.ServerSyncJob
        at com.squareup.moshi.Moshi$LookupChain.exceptionWithLookupStack(Moshi.java:349)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:150)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:98)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:68)
        at com.squareup.moshi.adapters.PolymorphicJsonAdapterFactory.create(PolymorphicJsonAdapterFactory.java:162)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:138)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:98)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:72)
        at com.pusher.pushnotifications.internal.ServerSyncHandler$Companion.obtain$pushnotifications_release(ServerSyncHandler.kt:77)
        at com.pusher.pushnotifications.PushNotificationsInstance$serverSyncHandler$1.invoke(PushNotificationsInstance.kt:106)
        at com.pusher.pushnotifications.PushNotificationsInstance$serverSyncHandler$1.invoke(PushNotificationsInstance.kt:93)
        at com.pusher.pushnotifications.PushNotificationsInstance.<init>(PushNotificationsInstance.kt:105)
        at com.pusher.pushnotifications.internal.PushNotificationsInitProvider.onCreate(ApplicationLifecycle.kt:67)
        at android.content.ContentProvider.attachInfo(ContentProvider.java:2092)
        at android.content.ContentProvider.attachInfo(ContentProvider.java:2066)
        at android.app.ActivityThread.installProvider(ActivityThread.java:6983)
        	... 10 more
     Caused by: java.lang.IllegalArgumentException: Cannot serialize Kotlin type com.pusher.pushnotifications.internal.StartJob. Reflective serialization of Kotlin classes without using kotlin-reflect has undefined and unexpected behavior. Please use KotlinJsonAdapter from the moshi-kotlin artifact or use code gen from the moshi-kotlin-codegen artifact.
        at com.squareup.moshi.ClassJsonAdapter$1.create(ClassJsonAdapter.java:83)
        at com.squareup.moshi.Moshi.adapter(Moshi.java:138)
        	... 24 more
```

Thank you

### PR DESCRIPTION
**Pull request by daniellevass**
## What
On of our dependencies, Moshi, changed how sealed types and our polymorphic way of reading data for our server sync events. 

From https://www.zacsweers.dev/a-closer-look-at-moshi-1-9/
> In projects using code gen, there are cases where Kotlin classes could have (appeared to) Just Work™️ before if you forgot to annotate them with @JsonClass. These will fail at runtime now.

We needed to follow the instructions Zac's written in https://github.com/ZacSweers/moshi-sealed to make our library work work in the new system. So I've added those dependencies and followed the new ways.

## Why
Users could experience a crash if they updated their own moshi dependency to 1.9.x - we shouldn't crash if they update. This should fix #100

### CODE REVIEW

File: pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt [hunk 1]
```diff
@@ -1,35 +1,44 @@
...
-                .withSubtype(SetUserIdJob::class.java, "SetUserIdJob")
-                .withSubtype(StopJob::class.java, "StopJob")
-    }
-}
+@JsonClass(generateAdapter = true, generator = "sealed:ServerSyncJob")
+sealed class ServerSyncJob: Serializable { }
 
 // If you add new fields to any of these data classes, make sure they have
 // default values otherwise they can be null when reading the old stored
 // JSON from disk.
```
Comment (marekoid) [inline pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt]: That comment have been no longer valid already before the changes in this PR but now it's a good opportunity to delete it.

File: pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt [hunk 2]
```diff
@@ -1,35 +1,44 @@
...
-                .withSubtype(RefreshTokenJob::class.java, "RefreshTokenJob")
-                .withSubtype(SubscribeJob::class.java, "SubscribeJob")
-                .withSubtype(UnsubscribeJob::class.java, "UnsubscribeJob")
-                .withSubtype(SetSubscriptionsJob::class.java, "SetSubscriptionsJob")
-                .withSubtype(ApplicationStartJob::class.java, "ApplicationStartJob")
-                .withSubtype(SetUserIdJob::class.java, "SetUserIdJob")
-                .withSubtype(StopJob::class.java, "StopJob")
-    }
-}
+@JsonClass(generateAdapter = true, generator = "sealed:ServerSyncJob")
```
Comment (marekoid) [inline pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt]: It does not seem like we're using code gen here (no kapt dependency). Those annotations here look redundant.

File: pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt [hunk 3]
```diff
@@ -1,35 +1,44 @@
...
+sealed class ServerSyncJob: Serializable { }
```
Comment (marekoid) [inline pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt]: `Serializable` marker looks redundant and if it is indeed redundant it is also confusing and error-prone.

File: sample_kotlin/build.gradle [hunk 4]
```diff
@@ -43,6 +43,8 @@ dependencies {
 
     implementation 'com.google.firebase:firebase-core:16.0.9'
     implementation 'com.google.firebase:firebase-messaging:18.0.0'
+
+    implementation 'com.squareup.moshi:moshi-kotlin:1.9.2'
```
Comment (marekoid) [inline sample_kotlin/build.gradle]: Why do we need to add moshi dependency to the sample_kotlin module?

File: pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt [hunk 5]
```diff
@@ -1,35 +1,44 @@
...
 
 // If you add new fields to any of these data classes, make sure they have
 // default values otherwise they can be null when reading the old stored
 // JSON from disk.
+@TypeLabel("StartJob")
+@JsonClass(generateAdapter = true)
 data class StartJob(val fcmToken: String, val knownPreviousClientIds: List<String>): ServerSyncJob()
```
Comment (marekoid) [inline pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt]: How about the migration? Won't we have a similar issue like the last time since we're changing how this sealed class hierarchy is unmarshalled? Won't there be any jobs persisted by a version prior to these changes of updating Moshi?

File: pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt (hunk 2 above)
Comment (marekoid) [reply pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt]: Or rather I hope that all that boiler-plate is redundant ;) Definitely the annotation values about generation seem redundant.

File: pushnotifications/build.gradle [hunk 6]
```diff
@@ -32,8 +32,10 @@ dependencies {
     implementation 'com.squareup.retrofit2:converter-gson:2.3.0'
     implementation 'com.squareup.retrofit2:retrofit:2.5.0'
     implementation 'com.squareup.tape2:tape:2.0.0-beta1'
-    implementation 'com.squareup.moshi:moshi:1.8.0'
-    implementation 'com.squareup.moshi:moshi-adapters:1.8.0'
+    implementation 'com.squareup.moshi:moshi-kotlin:1.9.2'
+    implementation 'com.squareup.moshi:moshi-adapters:1.9.2'
+    implementation "dev.zacsweers.moshisealed:moshi-sealed-reflect:0.1.0"
+    implementation "dev.zacsweers.moshisealed:moshi-sealed-annotations:0.1.0"
```
Comment (marekoid) [inline pushnotifications/build.gradle]: Not sure if the solution shouldn't be not to marshal a sealed class if it's not supported by Moshi... That extension lib created by one of the contributors is [WIP even when it comes to its API](https://www.reddit.com/r/androiddev/comments/dpjoy4/moshisealed_reflective_and_code_gen/f5vwsgn/). Also it seems like even with that lib, we need to remember about a lot of boiler-plate to put at each subclass of the sealed class.

Alternative solution could be about separately persisting given types of `ServerSyncJob`s so that we always know which type to expect when unmarshalling.

**Comment by daniellevass** [comment]
Marek left some feedback that we should double check that this works okay with people who already have a queue and will be upgrading, whether the core data itself would work. 

I thought we had tests that did this, and it turns out we do in `PersistentJobQueueTest` - we have three saved files that we access directly. However, they were failing 😱

Firstly, I've added a step in bitrise called `gradle test` which should run all our tests and actually notify us here if something has broken 🎉 

Secondly, it looks like the edge cases where the data models have changed is now different:

Previously if we had removed an attribute from an object that attribute became "null" or we needed to provide a default. Now that object will fail to parse :( 

Previously if we had added an attribute from an object, that attribute was ignored and the item was still returned. Now that object will also fail to parse :(

The current tests reflect the expected behaviour of using moshi 1.9 and have thusly been updated.

As the behaviour of this has changed I'm re-requesting reviews from you @luismfonseca -  I figure as we aren't likely to change the data models in the near term this shouldn't be an issue, but should be something you're aware of!

File: pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt (hunk 1 above)
Comment (luismfonseca) [reply pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt]: @daniellevass can you move some of your PR comment you did [here](https://github.com/pusher/push-notifications-android/pull/101#issuecomment-567611532) to this comment.

File: sample_kotlin/build.gradle (hunk 4 above)
Comment (luismfonseca) [reply sample_kotlin/build.gradle]: Yes, I'd expect things to work without this, so I think it's best if we don't include it here.

**Comment by daniellevass** [comment]
I updated the tests to more clearly explain what's expected and what happens when it fails to parse a record. The source of truth should be the tests and not this PR :-)

Coincidentally, I'd noticed that the saved files didn't contain the records that I expected. So I re-ran the uncommented code to re-create the files.

**Comment by marekoid** [comment]
Not directly related to this PR but it would be good to see test reports in Bitrise. That is what tests have been run to be sure all of them have been discovered.

File: pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt (hunk 1 above)
Comment (daniellevass) [reply pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt]: Have updated the comment to point people to the tests for the single source of truth! I updated the tests to more explicitly detail what results are expected and added comments there if data is expected not to be returned e.g. if a field is missing. Does that make sense?

File: pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt (hunk 2 above)
Comment (daniellevass) [reply pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt]: Yeah, I think you were right! I removed almost all of them and everything works correctly. I think the one on the sealed type is still necessary as it tells everything else it's a sealed type. Does that make sense?

File: pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt (hunk 3 above)
Comment (daniellevass) [reply pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt]: hmm. the code doesn't compile if I remove this. The ServerSyncHandler has a JobQueue ( which expects the types to be Serializable (https://github.com/pusher/push-notifications-android/blob/15f253d5ade2d783cd8aae52763de3c0bbcfcd11/pushnotifications/src/main/java/com/pusher/pushnotifications/internal/PersistentJobQueue.kt). I think this would require more refactoring to remove.

File: sample_kotlin/build.gradle (hunk 4 above)
Comment (daniellevass) [reply sample_kotlin/build.gradle]: I've removed this, it was useful to test that the demo app still worked correctly with this dependency (previously the app would crash).

File: pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt (hunk 5 above)
Comment (daniellevass) [reply pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt]: Yeah. We'll have difficulties migrating if any of the underlying classes change. As we've not changed any of the structure of those classes everything seems to be happy! I've updated the comment at the top of `ServerSyncJob` to explain what will happen if those classes change, and explained that the `ServerSyncJobTests` are our single source of truth for what's expected to happen :-)

File: pushnotifications/build.gradle (hunk 6 above)
Comment (daniellevass) [reply pushnotifications/build.gradle]: I think this is a question that @luismfonseca could help answer - are we okay using a WIP library to mitigate using sealed classes with moshi, or we could spend some time refactoring this piece of code to explicitly save the classes and not use a sealed type?

File: pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt (hunk 3 above)
Comment (marekoid) [reply pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt]: Oh I didn't expect serialisation is being done using two mechanisms. Thanks taking the feedback onboard and for checking! Yeah refactoring of that is not valuable enough to give it a shot.

File: pushnotifications/build.gradle (hunk 6 above)
Comment (marekoid) [reply pushnotifications/build.gradle]: From my side that comment is more to serve as lesson learned for new code, sorry for not expressing it clearly. I don't think it would be a good ROI in doing a broader and tricky refactoring here.

Comment (luismfonseca) [reply pushnotifications/build.gradle]: > are we okay using a WIP library to mitigate using sealed classes with moshi, or we could spend some time refactoring this piece of code to explicitly save the classes and not use a sealed type?

I think if we do sufficient testing on our end to make sure the library works well for us, and if we are happy with how it's performing, then we can use it. The fact that's not an official library means that we need to put in more effort to make sure it works for us.

Also, I think that ideally we have the discussions about the way to solve something should happen before a PR -- not always possible, but how we go about solving something should be discussed with the potential reviewers, minimizing the late "this is better if implemented in this completely different way".

Comment (daniellevass) [reply pushnotifications/build.gradle]: Yeah, that sounds fair! I'll test on the sample app again to be 100% certain this is safe for us to use.

Totally agree with discussing approaches before PRs - in future I'll set up a huddle before finalising my approach to discuss any hesitations anyone may have before starting the work. Does that work?

File: pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt (hunk 2 above)
Comment (marekoid) [reply pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt]: Awesome, thanks Danielle! I've checked that making the annotation that needs to stay not conveying information that code gen is involved (whilst it isn't), works (at least the tests involved in this diff are green on my machine).
```suggestion
@JsonClass(generateAdapter = false, generator = "sealed:ServerSyncJob")
```
Well it still sounds a bit like code gen is involved ("generator") but not that explicitly and hopefully `generateAdapter = false` clearly conveys that it is not. It feels though like an imperfection of that super early preview of the library extension so that's probably the best what we can achieve now.

File: pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt (hunk 5 above)
Comment (marekoid) [reply pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt]: Cool thanks for checking that!

File: pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt (hunk 2 above)
Comment (daniellevass) [reply pushnotifications/src/main/java/com/pusher/pushnotifications/internal/ServerSyncJob.kt]: That sounds fair! I've my code to reflect that.
